webdriver/tests/support/fixtures_bidi.py

```diff
@@ -905,3 +907,59 @@ def origin(protocol="https", domain="", subdomain=""):
         return urlunsplit((protocol, domain_value(domain, subdomain), "", "", ""))
 
     return origin
+
+
+@pytest_asyncio.fixture
+async def assert_file_dialog_canceled(bidi_session, inline, top_context):
+    async def assert_file_dialog_canceled(context=None):
+        context = context or top_context
+        cancel_event = await bidi_session.script.evaluate(
+            expression="""
+                new Promise(resolve => {
+                    const picker = document.createElement('input');
+                    picker.type = 'file';
+                    picker.addEventListener('cancel', (event) => {
+                        resolve(event.isTrusted);
+                    });
+                    picker.click();
+                })""",
+            target=ContextTarget(context["context"]),
+            await_promise=True,
+            user_activation=True
+        )
+
+        # Assert the `cancel` event is dispatched and the event is trusted.
+        assert cancel_event == {
+            'type': 'boolean',
+            'value': True
+        }
+
+    yield assert_file_dialog_canceled
+
+
+@pytest_asyncio.fixture
+async def assert_file_dialog_not_canceled(bidi_session, inline, top_context,
+        wait_for_future_safe):
+    async def assert_file_dialog_not_canceled(context=None):
+        context = context or top_context
+        cancel_event_future = asyncio.create_task(bidi_session.script.evaluate(
+            expression="""
+                        new Promise(resolve => {
+                            const picker = document.createElement('input');
+                            picker.type = 'file';
+                            picker.addEventListener('cancel', (event) => {
+                                resolve(event.isTrusted);
+                            });
+                            picker.click();
+                        })""",
+            target=ContextTarget(context["context"]),
+            await_promise=True,
+            user_activation=True
+        ))
+
+        with pytest.raises(TimeoutException):
+            await wait_for_future_safe(cancel_event_future, timeout=0.5)
+
+        cancel_event_future.cancel()
```

Comment:
What does it actually mean at this point when we keep the file picker open? In our case the main thread is blocked and as such as well all of our BiDi code. Are you able to close those pickers or do we have to force a restart of the browser after such a test?

Reply:
BTW this fixture was moved from `webdriver/tests/bidi/session/capabilities/unhandled_prompt_behavior/file/conftest.py`, where it did exactly the same thing.